feat: signing pkg for programmatic access

dtbuchholz opened the pull request:
## Summary

Adds a pkg for signing requests and a `sign` command, modeled off of [the signer CLI](https://github.com/tablelandnetwork/basin-cli/blob/bcalza/signer/cmd/signer/main.go) and [uploader.go](https://github.com/tablelandnetwork/basin-cli/blob/main/internal/app/uploader.go). Closes [ENG-787](https://linear.app/tableland/issue/ENG-787/allow-vaults-account-address-to-take-a-string).

## Details

It offers a few methods:

- `HexToECDSA`: Loads a private key from the given string and creates an ECDSA private key.
- `FileToECDSA`: Loads a private key from a file and creates an ECDSA private key.
- `NewSigner`: Creates a new signer with the given private key, provided by `LoadPrivateKey`.
- `SignFile`: Signs the given file with the signer and returns the signature as bytes.
- `SignBytes`: Signs the given bytes with the signer and returns the signature as bytes.
- `SignatureBytesToHex`: Converts bytes so they can be used in the URL POST request to write data to a vault.

## How it was tested

You can find the tests to ensure correct usage. Namely, I tested the CLI `signers` signature matched with the example I set up in the tests. I also created a working demo example here: https://github.com/dtbuchholz/demo-textile-http-api

brunocalza commented:
> @brunocalza i'm not sure if this is something we want to merge, but it was a request from IoTeX to make this available via a pkg instead of cmd. At the very least, they can access this branch.
> 
> If it's worthwhile pursuing, i can fix the linting issues.

In some sense, the `Signer` is already available for use as a package, just do

```
import "github.com/tablelandnetwork/basin-cli/internal/app"

signer := app.NewSigner(pk)
signer.SignFile("filename")
```

But it's a good idea to have its own package. But it would be good to refactor the CLI logic to use that package so we don't have 2 signers

dtbuchholz commented:
@brunocalza Okay gotcha, and I totally missed that impl in `internal/app`! 

Since this is sort of a one-off, partially for fun, I can take a stab at it. So, would it make sense for me to take your existing impl in [`uploader.go`](https://github.com/tablelandnetwork/basin-cli/blob/main/internal/app/uploader.go) for `NewSigner`, `Sign`, `SignFile`, `Sum`—plus, my new `LoadPrivateKey` method—and make those all available in `pkg/signing`? (They differ slightly from what I created in this PR.)

Then, do 2 things:
1) Import the `pkg/signing` lib into `uploader.go` and use the methods accordingly
2) Optionally, make your impl of the [`signer` CLI](https://github.com/tablelandnetwork/basin-cli/blob/bcalza/signer/cmd/signer/main.go) available as a new `sign` command in the `vaults` CLI? i.e., it should probably be an actual feature since it's related to the vaults functionality and useful if you're doing curl requests

brunocalza commented:
> @brunocalza Okay gotcha, and I totally missed that impl in `internal/app`!
> 
> Since this is sort of a one-off, partially for fun, I can take a stab at it. So, would it make sense for me to take your existing impl in [`uploader.go`](https://github.com/tablelandnetwork/basin-cli/blob/main/internal/app/uploader.go) for `NewSigner`, `Sign`, `SignFile`, `Sum`—plus, my new `LoadPrivateKey` method—and make those all available in `pkg/signing`? (They differ slightly from what I created in this PR.)
> 
> Then, do 2 things:
> 
> 1. Import the `pkg/signing` lib into `uploader.go` and use the methods accordingly
> 2. Optionally, make your impl of the [`signer` CLI](https://github.com/tablelandnetwork/basin-cli/blob/bcalza/signer/cmd/signer/main.go) available as a new `sign` command in the `vaults` CLI? i.e., it should probably be an actual feature since it's related to the vaults functionality and useful if you're doing curl requests

That approach makes sense to me. Let me know if you need help along the way.

dtbuchholz commented on pkg/signing/signing.go:
```diff
@@ -0,0 +1,130 @@
+
+// HexToECDSA parses a hex-encoded secp256k1 private key string to an ECDSA
+// private key.
+func HexToECDSA(hexKey string) (*ecdsa.PrivateKey, error) {
```
added a new method to create an ecdsa key from a private key string

dtbuchholz commented on pkg/signing/signing.go:
```diff
@@ -0,0 +1,130 @@
+
+// FileToECDSA parses a file path to a hex-encoded secp256k1 private key to an
+// ECDSA private key.
+func FileToECDSA(hexPath string) (*ecdsa.PrivateKey, error) {
```
added a new method to create an ecdsa key from a private key that's located at a file

dtbuchholz commented on pkg/signing/signing.go:
```diff
@@ -0,0 +1,130 @@
+package signing
```
here, the methods from `uploaded.go` have been moved to the `signing` package.

dtbuchholz commented on pkg/signing/signing.go:
```diff
@@ -0,0 +1,130 @@
+		_ = f.Close()
+	}()
+
+	// Check if the file is empty and return an error if it is
```
this is the only new piece compared to the old impl—since w3s will fail to write an empty file, i figured we should disable trying to upload files with empty content

dtbuchholz commented on pkg/signing/signing.go:
```diff
@@ -0,0 +1,130 @@
+}
+
+// SignBytes signs the provided bytes, returning the signature as a byte slice.
+func (s *Signer) SignBytes(data []byte) ([]byte, error) {
```
new method to sign bytes

dtbuchholz commented on pkg/signing/signing.go:
```diff
@@ -0,0 +1,130 @@
+}
+
+// SignatureBytesToHex converts a byte slice to a hex-encoded string.
+func SignatureBytesToHex(b []byte) string {
```
this is the only one i wasnt too sure about. i figured that it _might_ be useful for `SignFile` or `SignBytes` to return a bytes slice, which was the original functionality you had. then, you can take that slice and just convert it to a string signature, which is useful in the HTTP API.

but, i wasn't sure if keeping the return from `SignFile` or `SignBytes` as a bytes slice was useful. if not, i can remove `SignatureBytesToHex` and just make those methods return the signature as a string

brunocalza replied:
Well, in my view returning bytes is more "correct" in the sense that the only responsibility of the signing method is to sign. Whatever way you want to encode your bytes should be the responsibility of another part. The signing method has no opinions on how the signature is encoded. 

Not sure this method is really necessary since it's just calling the std lib method, the argument can be anything, not necessarily a signature

dtbuchholz commented on cmd/vaults/commands.go:
```diff
@@ -588,6 +589,50 @@ func newListEventsCommand() *cli.Command {
 	}
 }
 
+func newSignCommand() *cli.Command {
```
new command to `sign` data—i.e., `vaults sign --private-key abcd /path/to/file`

dtbuchholz commented on cmd/vaults/commands.go:
```diff
@@ -687,19 +732,28 @@ func newWalletCommand() *cli.Command {
-				Description: "The result of the `vaults account create` command will write a private key to a file, \n" +
-					"and this lets you retrieve the public key value for use in other commands.\n\n" +
-					"EXAMPLE:\n\nvaults account address /path/to/file",
+				UsageText: "vaults account address <file_path|hex_string>",
```
this is a backlog issue from the UX improvements project: [ENG-787](https://linear.app/tableland/issue/ENG-787/allow-vaults-account-address-to-take-a-string)

i.e., you can do `vaults account address /path/to/pk-file`, or you can do `vaults account address abcd1234`. i thought about adding a boolean flag like `--type file` or `--type string` in case that's a better UX, but it's a small feature

brunocalza replied:
yeah, i also prefer being explicit by using flags, but looks good

brunocalza commented on internal/app/uploader.go:
```diff
@@ -48,8 +43,9 @@ func (bu *VaultsUploader) Upload(
-	signature, err := signer.SignFile(filepath)
+	signer := signing.NewSigner(bu.privateKey)
+	signatureBytes, err := signer.SignFile(filepath)
+	signature := signing.SignatureBytesToHex(signatureBytes)
```
the err check should be above this line

brunocalza commented on pkg/signing/signing_test.go:
```diff
@@ -0,0 +1,210 @@
+	}
+}
+
+func TestPrivateKey(t *testing.T) {
```
Not sure but this test. `HexToECDSA` and `FileToECDSA` are just calling std lib methods that are well-tested. This test does not add more signal. Maybe just the happy path test is good enough because it tests that the method calls the correct std library method. But all the other error tests are kind of redundant, i guess

brunocalza reviewed (approved):
looks pretty good 💯 left some comments but no big changes really required

dtbuchholz commented:
@brunocalza i agree with all of your points! i've done the following:
- Removed the `SignatureBytesToHex` method
- Removed failure tests for `HexToECDSA` and `FileToECDSA`
- Fixed order of err check in `uploader.go`
- Slight adjustment to `vaults account address` where it defaults to a filepath, but you can also do `vaults account address --string abcd1234` such that you can explicitly specify it's a string. Thus, backward compatible and clearer on intended behavior.
